ambry-account/src/main/java/com/github/ambry/account/mysql/MySqlDataAccessor.java

```diff
@@ -54,63 +64,163 @@
+      closeActiveConnection();
+      activeConnection = null;
+      connectedEndpoint = null;
+    }
+
+    // If the active connection is good and it's the best endpoint, keep it.
+    if (activeConnection != null && !isBetterEndpoint(sortedDbEndpoints.get(0), connectedEndpoint)) {
       return activeConnection;
     }
-    if (activeConnection != null) {
-      activeConnection.close();
+    // See if we can do better
+    Pair<DbEndpoint, Connection> endpointConnectionPair = connectToBestAvailableEndpoint(needWritable);
+    if (connectedEndpoint == endpointConnectionPair.getFirst()) {
+      // No better endpoint  was available.
+      logger.debug("Still connected to {}", connectedEndpoint.getUrl());
+    } else {
+      // New connection established
+      connectedEndpoint = endpointConnectionPair.getFirst();
+      String qualifier = connectedEndpoint.isWriteable() ? "writable" : "read-only";
+      logger.info("Connected to {} enpoint: {}", qualifier, connectedEndpoint.getUrl());
+      closeActiveConnection();
+      activeConnection = endpointConnectionPair.getSecond();
     }
-    Properties credentials = new Properties();
-    credentials.setProperty("user", mysqlUser);
-    credentials.setProperty("password", mysqlPassword);
-    activeConnection = mysqlDriver.connect(mysqlUrl, credentials);
     return activeConnection;
   }
 
+  /**
+   * Connect to the best available database instance that matches the specified criteria.<br/>
+   * Order of preference for instances to connect to:
+   * <OL>
+   *    <LI/> Writeable instance in local colo
+   *    <LI/> Writable instance in any colo
+   *    <LI/> Read-only instance in local colo (if needWritable is false)
+   *    <LI/> Read-only instance in any colo (if needWritable is false)
+   *  </OL>
+   * @param needWritable whether the endpoint needs to be writable
+   * @return a pair of {@link DbEndpoint} and corresponding {@link Connection}.
+   * @throws SQLException if connection could not be made to a suitable endpoint.
+   */
+  private Pair<DbEndpoint, Connection> connectToBestAvailableEndpoint(boolean needWritable) throws SQLException {
+    SQLException lastException = null;
+    for (DbEndpoint candidateEndpoint : sortedDbEndpoints) {
+      if (!isBetterEndpoint(candidateEndpoint, connectedEndpoint)) {
+        // What we have is the best we can do.  Is it good enough?
+        if (needWritable && !connectedEndpoint.isWriteable()) {
+          throw Optional.of(lastException).orElse(noWritableEndpointException);
+        } else {
+          return new Pair<>(connectedEndpoint, activeConnection);
+        }
+      }
+      if (needWritable && !candidateEndpoint.isWriteable()) {
+        throw Optional.of(lastException).orElse(noWritableEndpointException);
+      }
+
+      // Attempt to connect to candidate endpoint
+      Properties credentials = new Properties();
+      credentials.setProperty("user", candidateEndpoint.getUsername());
```

Comment:
I wonder if there are constants for the mysql driver properties anywhere.

Reply:
I would think so but didn't see any.